File: backend/app/schemas/finance.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import TransactionType


class ChildBalanceResponse(BaseModel):
    child_id: int
    child_name: str
    balance_cents: int


class TransactionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    child_id: int
    amount_cents: int
    type: TransactionType
    memo: str
    created_at: datetime


class CreateTransactionRequest(BaseModel):
    child_id: int = Field(gt=0)
    amount_cents: int = Field(ge=-10_000_000, le=10_000_000)
    type: TransactionType
    memo: str = Field(default="", max_length=500)

    @model_validator(mode="after")
    def validate_manual_transaction(self) -> "CreateTransactionRequest":
        if self.type == TransactionType.CHORE_APPROVAL:
            raise ValueError("Chore approval transactions are created by the approval workflow.")
        if self.amount_cents == 0:
            raise ValueError("Transaction amount cannot be zero.")
        return self

File: packages/family-api/src/models/finance.ts
export type ChoreTransactionType =
  | "CHORE_APPROVAL"
  | "BONUS"
  | "PAYMENT"
  | "ADJUSTMENT";

export interface ChildBalance {
  child_id: number;
  child_name: string;
  balance_cents: number;
}

export interface ChoreTransaction {
  id: number;
  child_id: number;
  amount_cents: number;
  type: ChoreTransactionType;
  memo: string;
  created_at: string;
}

export interface CreateChoreTransactionRequest {
  child_id: number;
  amount_cents: number;
  type: Exclude<ChoreTransactionType, "CHORE_APPROVAL">;
  memo?: string;
}
